genie-docs/src/docs/asciidoc/features/_clusterLoadBalancing.adoc

```diff
@@ -0,0 +1,192 @@
+store them in an invocation order list. Meaning that when multiple clusters are selected from the database based on tags
+Genie will send the list of clusters to the implementations in preference order until one of them returns the cluster to
+use.
+
+Genie currently ships with two implementations of this interface which are described below.
+
+==== RandomizedClusterLoadBalancerImpl
+
+As the name indicates this load balancer simply uses a random number generator to select a cluster from the list by
+index. There is no intelligence to this load balancer algorithm but it does provide very close to an equal distribution
+between clusters if the same tags are always used.
+
+This implementation is the "default" implementation. It has the lowest priority order so if all other active
+implementations fail to select a cluster this implementation will be invoked and chose randomly.
+
+==== ScriptLoadBalancer
+
+This implementation, first introduced in 3.1.0, allows administrators to provide a script to be invoked at runtime to
+decide which cluster to select. Currently `JavaScript` and `Groovy` are supported out of the box but others (like
+`Python`, `Ruby`, etc) could be supported by adding their implementations of the
+https://docs.oracle.com/javase/8/docs/api/javax/script/ScriptEngine.html[ScriptEngine] interface to the Genie classpath.
+
+===== Configuration
+
+The script load balancer is disabled by default. To enable it an admin must set the property
+`genie.jobs.clusters.loadBalancers.script.enabled` to true.
+
+Other properties:
+
+|===
+|Property |Description| Default Value
+
+|genie.jobs.clusters.loadBalancers.script.destination
+|The location on disk where the script source file should be stored after it is downloaded from
+`genie.jobs.clusters.loadBalancers.script.source`. The file will be given the same name.
+|file:///tmp/genie/loadbalancers/script/destination/
+
+|genie.jobs.clusters.loadBalancers.script.enabled
+|Whether the script based load balancer should be enabled for the system or not.
+See also: `genie.jobs.clusters.loadBalancers.script.source`
+See also: `genie.jobs.clusters.loadBalancers.script.destination`
+|false
+
+|genie.jobs.clusters.loadBalancers.script.refreshRate
+|How frequently to refresh the load balancer script (in milliseconds)
+|300000
+
+|genie.jobs.clusters.loadBalancers.script.source
+|The location of the script the load balancer should load to evaluate which cluster to use for a job request
+|file:///tmp/genie/loadBalancers/script/source/loadBalance.js
+
+|genie.jobs.clusters.loadBalancers.script.timeout
+|The amount of time (in milliseconds) that the system will attempt to run the cluster load balancer script before it
+forces a timeout
+|5000
+
+|===
+
+===== Script Contract
+
+The contract between the script and the Java code is as follows:
+
+.Script Inputs
+|===
+|Parameter |Description
+
+|clusters
+|JSON array of cluster objects (serialized as a string) of the clusters to be evaluated
```

Comment:
As of now, this is guaranteed to be non-empty. Should we make it explicit here and in Javadoc?